### ISSUE TITLE
a problem about training time

### ISSUE THREAD

**Issue by lgxccc** [dki-lab/ArcaneQA#5]
Dear Gu,

Thank you very much for your invaluable assistance in answering my previous questions. 

I am currently utilizing an RTX 4070 graphics card, 32GB of RAM, and an i5-12600KF processor for my project. I have configured my gradient accumulation steps to 4. However, I have encountered a significant time constraint during training. Specifically, it has taken me approximately 4 days to complete the first epoch, reaching only 50% progress.

I am curious to know if you also experience similar training times with your setup. It would be tremendously helpful to understand if such extended training durations are common or if there are potential strategies to reduce the training time.

Thank you once again for your exceptional contributions and expertise. I am looking forward to your valuable insights and suggestions.

Best regards

**Comment by entslscheia** [comment]
In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet.
4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?

**Comment by lgxccc** [comment]
> In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet. 4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?

Yes, indeed, the GPU utilization during model training is consistently around 0%. I would like to inquire about how to use the SPARQL engine and how to utilize the SPARQL cache. Could you please provide some guidance on these topics? I appreciate your assistance. Thank you.

**Comment by lgxccc** [comment]
> In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet. 4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?

I apologize for my negligence in not adjusting the path for the SPARQL cache file. I have now corrected the path, and the SPARQL cache is functioning properly. Additionally, I expect the first epoch to complete training in approximately 40 hours. However, I have noticed that the GPU utilization remains around 0% to 10%. Could you please clarify if this is normal behavior or if it could be due to my failure to utilize the SPARQL engine? I appreciate your assistance and apologize for any inconvenience caused. Thank you.

**Comment by entslscheia** [comment]
> > In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet. 4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?
> 
> I apologize for my negligence in not adjusting the path for the SPARQL cache file. I have now corrected the path, and the SPARQL cache is functioning properly. Additionally, I expect the first epoch to complete training in approximately 40 hours. However, I have noticed that the GPU utilization remains around 0% to 10%. Could you please clarify if this is normal behavior or if it could be due to my failure to utilize the SPARQL engine? I appreciate your assistance and apologize for any inconvenience caused. Thank you.

Low volatile rate basically means the process is occupied with CPU jobs (mostly executing SPARQL queries in this case), but I think you should expect the volatile rate to go up as training continues (because more queries will be cached and thus obviate the need to re-query the SPARQL endpoint).

Generally, there might be two tips to make SPARQL query execution faster: 1) set up the virtuoso endpoint on the same machine when training your model (setting up it on a remote machine might lead to some communication overhead), and 2) make sure your server have enough memory.

It's my pleasure to help. No inconvenience at all! Feel free to reach out if you have any further concerns :)

**Comment by lgxccc** [comment]
> > > In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet. 4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?
> > 
> > 
> > I apologize for my negligence in not adjusting the path for the SPARQL cache file. I have now corrected the path, and the SPARQL cache is functioning properly. Additionally, I expect the first epoch to complete training in approximately 40 hours. However, I have noticed that the GPU utilization remains around 0% to 10%. Could you please clarify if this is normal behavior or if it could be due to my failure to utilize the SPARQL engine? I appreciate your assistance and apologize for any inconvenience caused. Thank you.
> 
> Low volatile rate basically means the process is occupied with CPU jobs (mostly executing SPARQL queries in this case), but I think you should expect the volatile rate to go up as training continues (because more queries will be cached and thus obviate the need to re-query the SPARQL endpoint).
> 
> Generally, there might be two tips to make SPARQL query execution faster: 1) set up the virtuoso endpoint on the same machine when training your model (setting up it on a remote machine might lead to some communication overhead), and 2) make sure your server have enough memory.
> 
> It's my pleasure to help. No inconvenience at all! Feel free to reach out if you have any further concerns :)

With your help, I successfully reduced the training time to one day, but the testing time is still very slow, about 10 days. I found that the testing needs to use 'bundle search' that was not used during training, which makes the testing speed very slow. May I ask if this is also a problem with 'sparql cache'

**Comment by lgxccc** [comment]
> > > In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet. 4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?
> > 
> > 
> > I apologize for my negligence in not adjusting the path for the SPARQL cache file. I have now corrected the path, and the SPARQL cache is functioning properly. Additionally, I expect the first epoch to complete training in approximately 40 hours. However, I have noticed that the GPU utilization remains around 0% to 10%. Could you please clarify if this is normal behavior or if it could be due to my failure to utilize the SPARQL engine? I appreciate your assistance and apologize for any inconvenience caused. Thank you.
> 
> Low volatile rate basically means the process is occupied with CPU jobs (mostly executing SPARQL queries in this case), but I think you should expect the volatile rate to go up as training continues (because more queries will be cached and thus obviate the need to re-query the SPARQL endpoint).
> 
> Generally, there might be two tips to make SPARQL query execution faster: 1) set up the virtuoso endpoint on the same machine when training your model (setting up it on a remote machine might lead to some communication overhead), and 2) make sure your server have enough memory.
> 
> It's my pleasure to help. No inconvenience at all! Feel free to reach out if you have any further concerns :)

With your help, I successfully reduced the training time to one day, but the testing time is still very slow, about 10 days. I found that the testing needs to use 'beam search' that was not used during training, which makes the testing speed very slow. May I ask if this is also a problem with 'sparql cache'

**Comment by lgxccc** [comment]
> > > In my case, it took around 12 hrs to train one epoch with the SPARQL cache. It might take a bit longer (i.e., around 20 hrs) for the first epoch because many queries are not cached yet. 4 days one epoch is just way too long. Maybe it's related to your SPARQL engine? Could you check the volatile rate of your GPU? Is it 0% for most of the time?
> > 
> > 
> > I apologize for my negligence in not adjusting the path for the SPARQL cache file. I have now corrected the path, and the SPARQL cache is functioning properly. Additionally, I expect the first epoch to complete training in approximately 40 hours. However, I have noticed that the GPU utilization remains around 0% to 10%. Could you please clarify if this is normal behavior or if it could be due to my failure to utilize the SPARQL engine? I appreciate your assistance and apologize for any inconvenience caused. Thank you.
> 
> Low volatile rate basically means the process is occupied with CPU jobs (mostly executing SPARQL queries in this case), but I think you should expect the volatile rate to go up as training continues (because more queries will be cached and thus obviate the need to re-query the SPARQL endpoint).
> 
> Generally, there might be two tips to make SPARQL query execution faster: 1) set up the virtuoso endpoint on the same machine when training your model (setting up it on a remote machine might lead to some communication overhead), and 2) make sure your server have enough memory.
> 
> It's my pleasure to help. No inconvenience at all! Feel free to reach out if you have any further concerns :)

When testing the model, I ran the code that was not executed during training. One particular line of code took a long time:
`predictions = self._forward_beam_search(state)`
which caused the model to take more than ten days to run on the test set.
I'm wondering if this code is also related to 'sparql_cache'? I'm conducting a model-agnostic experiment, and I just need to perform the final testing step. Brother, please help me out. Bringing you so much trouble,  I greatly appreciate your help!

**Comment by entslscheia** [comment]
Hi @lgxccc ,

Sorry for the late response.
In my experiments, I think it takes around 10 hrs to finish the prediction on the entire test set (i.e., ~13,000 questions) with sparql cache on. Actually, the behavior for inference is the almost the same as the behavior for validation during training, which means `predictions = self._forward_beam_search(state)` is also executed in validation. I'd suggest you reuse the same cache file you obtained during training, which might cover many of the SPARQL queries the model encounters during inference. Let me know if you have got it fixed!

As I've mentioned earlier, it's our pleasure to help you with using our model. Feel free to reach out if you have any further concerns.
